### PR TITLE
LPD-32094 - ERC Support for Postal Addresses

### PR DESCRIPTION
**Pull request by brbalazs**
https://liferay.atlassian.net/browse/LPD-32094

### CODE REVIEW

**Comment by brbalazs** [comment]
ci:test:sf

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/rest-openapi.yaml [hunk 1]
```diff
@@ -4582,6 +4586,99 @@ paths:
...
+            responses:
+                204:
+                    content:
+                        application/json: {}
+                        application/xml: {}
+            tags: ["PostalAddress"]
+        get:
+            description:
+                Retrieves the postal address using external reference code.
+            operationId: getPostalAddressByExternalReferenceCodey
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/rest-openapi.yaml:4609]: there is a y at the end

File: portal-impl/src/com/liferay/portal/service/impl/AddressLocalServiceImpl.java [hunk 2]
```diff
@@ -312,6 +313,22 @@ public Address updateAddress(
 		return address;
 	}
 
+	@Override
+	public Address updateAddressExternalReferenceCode(
```
Comment (smottal) [inline portal-impl/src/com/liferay/portal/service/impl/AddressLocalServiceImpl.java:317]: can we call this only `updateExternalReferenceCode`?

File: portal-impl/src/com/liferay/portal/service/impl/AddressServiceImpl.java [hunk 3]
```diff
@@ -138,4 +138,17 @@ public Address updateAddress(
 			regionId, countryId, listTypeId, mailing, primary, phoneNumber);
 	}
 
+	@Override
+	public Address updateAddressExternalReferenceCode(
```
Comment (smottal) [inline portal-impl/src/com/liferay/portal/service/impl/AddressServiceImpl.java:142]: can we call this only `updateExternalReferenceCode`?

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 4]
```diff
@@ -61,6 +64,25 @@ public void deletePostalAddress(Long postalAddressId) throws Exception {
 		_addressService.deleteAddress(postalAddressId);
 	}
 
+	@Override
+	public void deletePostalAddressByExternalReferenceCode(
+			String postalAddressExternalReferenceCode)
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:69]: is it called like this also in the "base"? I think should be only `externalReferenceCode`

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 5]
```diff
@@ -61,6 +64,25 @@ public void deletePostalAddress(Long postalAddressId) throws Exception {
...
+		throws Exception {
+
+		Address address =
+			_addressLocalService.getAddressByExternalReferenceCode(
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:73]: I'm crying for this 😢  ... not your fault but it's wrong ... it should be a fetch, not a get.
We need to ask to @riccardo-alberti what we want to do... change "everything" or cross finger

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 6]
```diff
@@ -61,6 +64,25 @@ public void deletePostalAddress(Long postalAddressId) throws Exception {
...
+				postalAddressExternalReferenceCode,
+				contextCompany.getCompanyId());
+
+		if (address == null) {
+			throw new NoSuchAddressException(
+				"No Address found with external reference code " +
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:79]: can you just check if in other places we put the name of the model (like `Address` here lowercase or not?

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 7]
```diff
@@ -130,6 +152,25 @@ public PostalAddress getPostalAddress(Long postalAddressId)
 			contextAcceptLanguage.getPreferredLocale());
 	}
 
+	@Override
+	public PostalAddress getPostalAddressByExternalReferenceCode(
+			String postalAddressExternalReferenceCode)
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:157]: also the name here I think it's wrong

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 8]
```diff
@@ -130,6 +152,25 @@ public PostalAddress getPostalAddress(Long postalAddressId)
...
+		throws Exception {
+
+		Address address =
+			_addressLocalService.getAddressByExternalReferenceCode(
+				postalAddressExternalReferenceCode,
+				contextCompany.getCompanyId());
+
+		if (address == null) {
+			throw new NoSuchAddressException(
+				"No Address found with external reference code " +
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:167]: same A -> a here ... to be checked

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 9]
```diff
@@ -231,12 +272,38 @@ public PostalAddress patchPostalAddress(
...
+
 		return PostalAddressUtil.toPostalAddress(
 			contextAcceptLanguage.isAcceptAllLanguages(), address,
 			contextCompany.getCompanyId(),
 			contextAcceptLanguage.getPreferredLocale());
 	}
 
+	@Override
+	public PostalAddress patchPostalAddressByExternalReferenceCode(
+			String postalAddressExternalReferenceCode,
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:289]: same here

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 10]
```diff
@@ -231,12 +272,38 @@ public PostalAddress patchPostalAddress(
...
+		throws Exception {
+
+		Address address =
+			_addressLocalService.getAddressByExternalReferenceCode(
+				postalAddressExternalReferenceCode,
+				contextCompany.getCompanyId());
+
+		if (address == null) {
+			throw new NoSuchAddressException(
+				"No Address found with external reference code " +
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:300]: A -> a

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 11]
```diff
@@ -288,12 +355,38 @@ public PostalAddress putPostalAddress(
...
+
 		return PostalAddressUtil.toPostalAddress(
 			contextAcceptLanguage.isAcceptAllLanguages(), address,
 			contextCompany.getCompanyId(),
 			contextAcceptLanguage.getPreferredLocale());
 	}
 
+	@Override
+	public PostalAddress putPostalAddressByExternalReferenceCode(
+			String postalAddressExternalReferenceCode,
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:372]: same

File: modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java [hunk 12]
```diff
@@ -288,12 +355,38 @@ public PostalAddress putPostalAddress(
...
+		throws Exception {
+
+		Address address =
+			_addressLocalService.getAddressByExternalReferenceCode(
+				postalAddressExternalReferenceCode,
+				contextCompany.getCompanyId());
+
+		if (address == null) {
+			throw new NoSuchAddressException(
+				"No Address found with external reference code " +
```
Comment (smottal) [inline modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/PostalAddressResourceImpl.java:383]: same

**Review by smottal — CHANGES_REQUESTED** [review]

**Comment by brbalazs** [comment]
ci:test:sf

File: portal-impl/src/com/liferay/portal/service/impl/AddressServiceImpl.java [hunk 13]
```diff
@@ -138,4 +138,16 @@ public Address updateAddress(
...
 	}
 
+	public Address updateExternalReferenceCode(
+			Address address, String externalReferenceCode)
+		throws PortalException {
+
+		CommonPermissionUtil.check(
+			getPermissionChecker(), address.getClassNameId(),
+			address.getClassPK(), ActionKeys.UPDATE);
+
```
Comment (smottal) [inline portal-impl/src/com/liferay/portal/service/impl/AddressServiceImpl.java:148]: I think this is not enough compared with the updateAddress method where we check another permission if it's an address related to an account

**Review by smottal — CHANGES_REQUESTED** [review]

**Comment by brbalazs** [comment]
ci:test:sf

**Review by smottal — APPROVED** [review]

**Comment by brbalazs** [comment]
ci:test:relevant

**Comment by smottal** [comment]
ci:forward

**Comment by smottal** [comment]
ci:report:51747136

**Comment by smottal** [comment]
ci:forward:force

**Pull request closed without merging**